File: blind75_grind75/ContainerWithMostWater.java
package grind75;

public class ContainerWithMostWater {

    public int maxArea1(int[] height) {
        int max = 0;
        int head = 0, tail = height.length - 1;
        while (head < tail) {
            int h = Math.min(height[head], height[tail]);
            max = Math.max(max, (tail - head) * h);
            if (height[head] > height[tail]) {
                tail--;
            } else {
                head++;
            }
        }
        return max;
    }

    public int maxArea(int[] height) {
        int i = 0, j = height.length - 1 ;
        int maxLeft, maxRight;
        int area = 0, maxArea = 0 ;

        while (i < j) {
            maxLeft = height[i] ;
            maxRight = height[j] ;

            if (maxRight > maxLeft) area = maxLeft * (j-i) ;
            else area = maxRight * (j-i) ;

            maxArea = Math.max(maxArea, area);

            if (maxRight > maxLeft) {
                while (i < j && height[i] <= maxLeft) {
                    i++ ;
                }
            } else {
                while (i < j && height[j] <= maxRight) {
                    j-- ;
                }
            }
        }
        return maxArea ;
    }
}

File: go/grind75/container_with_most_water.go
package grind75

func maxArea(height []int) int {
	left, right := 0, len(height)-1
	maxArea := 0
	for left < right {
		width := right-left
		high := min(height[left], height[right])
		maxArea = max(maxArea, width*high)
		if (height[left] > height[right]) {
			right--
		} else {
			left++
		}
	}
	return maxArea
}

// func min(a, b int) int {
// 	if a > b {
// 		return b
// 	}
// 	return a
// }
func Max(x, y int)int{
    if x >= y {
        return x
    }
    
    return y
}

func maxArea2(height []int) int {
    
    // length of input array
    size := len(height)
    
    // tow pointers, left init as 0, right init as size-1
    left, right := 0, size-1
    
    // maximal width between leftmost stick and rightmost stick
    maxWidth := size-1
    
    // area also known as the amount of water
    area := 0
    
    // trade-off between width and height
    // scan each possible width and compute maximal area
    for width := maxWidth ; width > 0 ; width-- {
        
        if height[left] < height[right]{
            
            // the height of lefthand side is shorter
            area = Max(area, width * height[left])
            
            // update left index to righthand side
            left += 1
            
        }else{
            
            // the height of righthand side is shorter
            area = Max(area, width * height[right])
            
            // update left index to righthand side
            right -= 1
        }
        
    }
    return area
}